### PR TITLE
[WIP] support k2 ctc

### PR DESCRIPTION
**Pull request by yaguanghu**
This is an initial PR for k2 ctc loss support in ESPnet. I'm running running experiments on Aishell datasets, and the loss seems normal.
I will also verify it in espnet2.

### CODE REVIEW

**Comment by sw005320** [comment]
Added @brianyan918 

@yaguanghu, many thanks!
What I'm concerning is how we establish robust k2 installation in the espnet environment.
warpctc had several issues and we make it in the current form.
It would be great if you help us in this regard.

File: espnet/nets/pytorch_backend/ctc.py [hunk 1]
```diff
@@ -25,13 +25,14 @@ def __init__(self, odim, eprojs, dropout_rate, ctc_type="warpctc", reduce=True):
...
-        self.ctc_type = (
-            ctc_type
-            if LooseVersion(torch.__version__) < LooseVersion("1.7.0")
-            else "builtin"
-        )
+        # self.ctc_type = (
+        #     ctc_type
+        #     if LooseVersion(torch.__version__) < LooseVersion("1.7.0")
+        #     else "builtin"
+        # )
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/ctc.py:35]: I think this part should be refined.
This part was made since warpctc does not support pytorch 1.7.0
gtn and k2 should work with Pytorch >= 1.7.0 while they would not work with some older Pytorch versions.
It would be great if you refine this condition and adding gtn and k2 unit test so that it could pass the CI check (we're checking PyTorch 1.0.1, 1.1.0, 1.2.0, 1.3.1, 1.4.0, 1.5.1, 1.6.0, and 1.7.1)

Comment (yaguanghu) [reply espnet/nets/pytorch_backend/ctc.py:35]: I will try to refine this logic.

**Comment by yaguanghu** [comment]
> Added @brianyan918
> 
> @yaguanghu, many thanks!
> What I'm concerning is how we establish robust k2 installation in the espnet environment.
> warpctc had several issues and we make it in the current form.
> It would be great if you help us in this regard.

K2 team members are trying to improve the installation. Hopefully we just need to run `pip install k2`.

**Comment by danpovey** [comment]
Yes, should just be installable with pip, but the version to be installed depends on the PyTorch version.
I believe we support 1.6 and 1.7 (not sure the precise version numbers within that).
@csukuangfj should be able to comment.

**Comment by csukuangfj** [comment]
k2 can be installed by just running
```bash
pip install k2
```
or
```bash
pip install --pre k2 # the latest version on PyPI, not necessarily the same with the latest master
```

k2 installed with pip supports:
- Python 3.6, 3.7, 3.8
- CUDA 10.1
- PyTorch 1.7.1

---

If k2 is installed from source, it is known to work with
- Python 3.6, 3.7, 3.8 (the minimum required version is 3.6)
- CUDA 10.1, 10.2, 11.0 (other CUDA versions may also work, but have not been tested)
- PyTorch 1.6.x, 1.7.x (1.6.0, 1.7.0, 1.7.1 are known to work; other versions may also work, but have not been tested)

We have a detailed documentation about the installation of k2. Please see
https://k2.readthedocs.io/en/latest/installation.html

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            else:
+                arcs += f'{i} {j} {tokens[j]} {tokens[j]} 0.0\n'
+        arcs += f'{i} {final_state} -1 -1 0.0\n'
+    arcs += f'{final_state}'
+    ans = k2.Fsa.from_str(arcs)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+class CtcTrainingGraphCompiler(object):
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:89]: We have another version of building graphs where all operations are moved to CUDA.
Please see 
https://github.com/k2-fsa/snowfall/blob/2e679045be6fc251e9fbb5fff614a5e35104136f/snowfall/training/mmi_mbr_graph.py#L109

Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:89]: There may be problems with multi-gpu training if we build graphs with CUDA and put graphs in a specific GPU memories. I haven't figured out how to deal with this problem, but I don't think it's a problem because of the speed of building graphs is quite faster and the size of CTC graph is small enough.

Comment (danpovey) [reply espnet/nets/pytorch_backend/k2_ctc.py:89]: mm I didn't like setting the gpu_id explicitly either.  I feel like we should be getting it from somewhere else.

Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:95]: Anyway, I have implemented the CUDA graphs building method. But it will not be used until I figure out how to work with multi-gpu training.

**Comment by sw005320** [comment]
Please fix the CI error.
Check https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md if you want to test the CI test locally.

File: espnet/nets/pytorch_backend/ctc.py [hunk 3]
```diff
@@ -45,6 +46,12 @@ def __init__(self, odim, eprojs, dropout_rate, ctc_type="warpctc", reduce=True):
...
             self.ctc_loss = GTNCTCLossFunction.apply
+        elif self.ctc_type == "k2ctc":
+            logging.info(f"CTC type is {self.ctc_type}")
+            reduction_type = "sum" if reduce else "none"
+            from espnet.nets.pytorch_backend.k2_ctc import K2CTCLoss
+
+            self.ctc_loss = K2CTCLoss(odim, reduction=reduction_type)
         else:
             raise ValueError(
                 'ctc_type must be "builtin" or "warpctc": {}'.format(self.ctc_type)
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/ctc.py:57]: I think this error message has to be updated to mention `gtnctc` and `k2ctc`.

File: test/espnet2/asr/test_k2.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    pytorch_loss = pytroch_ctc_loss(
+        pytorch_log_probs, target, input_lengths, target_lengths
+    )
+    pytorch_loss.backward()
+
+    devices = [torch.device("cpu")]
+    if torch.cuda.is_available():
+        devices.append(torch.device("cuda"))
+    for device in devices:
+        k2_ctc_loss = K2CTCLoss(C)
```
Comment (csukuangfj) [inline test/espnet2/asr/test_k2.py:35]: K2CTCLoss supports two extra arguments: `reduction` and `device`. It would be great if there are tests covering these two arguments too.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            logging.error(
+                f"k2 ctc loss reduction type: {reduction} is not supported yet,"
+                "change it to be 'sum'"
+            )
+            self.reduction = "sum"
+        self.graph_compiler = CtcTrainingGraphCompiler(odim, device=device)
+
+    def forward(
+        self,
+        log_probs: torch.Tensor,
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:36]: Can you also add documentation to these arguments?

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+        decoding_graph = self.graph_compiler.compile(targets, target_lengths)
+        if self.device != log_probs.device:
+            decoding_graph = decoding_graph.to(log_probs.device)
+
+        target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 100.0)
+        tot_scores = target_graph.get_tot_scores(
+            log_semiring=True, use_double_scores=False
+        )
+        if self.reduction == "none":
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:63]: What are the differences between `none` and `sum`?
Did you forget to use an early `return` inside the `if` statement?

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    ans = k2.Fsa.from_str(arcs)
+    ans = k2.arc_sort(ans)
+    return ans
+
+
+class CtcTrainingGraphCompiler(object):
+    def __init__(self, odim: int, device: torch.device):
+        """
+        Args:
+        odim:
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:107]: The indentation of the docstring for the argument is not correct.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        decoding_graph = k2.intersect(
+            self.ctc_topo_inv, label_graph, treat_epsilons_specially=False
+        )
+        decoding_graph = k2.connect(decoding_graph.invert_())
+        return decoding_graph
+
+
+def get_tot_objf_and_num_frames(
+    tot_scores: torch.Tensor, frames_per_seq: torch.Tensor
+) -> Tuple[float, int, int]:
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:147]: The type hints for the return values are not correct. The three values are tensors, not scalars.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+        dense_fsa_vec = k2.DenseFsaVec(log_probs, supervision_segments)
+
+        decoding_graph = self.graph_compiler.compile(targets, target_lengths)
+        if self.device != log_probs.device:
+            decoding_graph = decoding_graph.to(log_probs.device)
+
+        target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 100.0)
+        tot_scores = target_graph.get_tot_scores(
+            log_semiring=True, use_double_scores=False
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:61]: I would recommend setting `use_double_scores` to True.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+            [
+                self.compile_one_and_cache(
+                    texts[texts_end_index[i] : texts_end_index[i + 1]]
+                )
+                for i in range(texts_lengths.shape[0] - 1)
+            ]
+        )
+
+        # make sure the gradient is not accumulated
+        decoding_graphs.requires_grad_(False)
```
[the remark this answers is not in the document]
Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:131]: `intersect with ctc_topo` is already done in `compile_one_and_cache` function.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        if self.device != log_probs.device:
+            decoding_graph = decoding_graph.to(log_probs.device)
+
+        target_graph = k2.intersect_dense(decoding_graph, dense_fsa_vec, 100.0)
+        tot_scores = target_graph.get_tot_scores(
+            log_semiring=True, use_double_scores=False
+        )
+        if self.reduction == "none":
+            tot_scores = tot_scores / target_lengths.to(log_probs.device)
+        (tot_score, tot_frames, all_frames) = get_tot_objf_and_num_frames(
```
[the remark this answers is not in the document]
Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:65]: I think this is only for debugging, but anyway it's not used in current code.

File: espnet/nets/pytorch_backend/k2_ctc.py (hunk 6 above)
Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:63]: These two different types are just different loss normalization method. There is no logic problem.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        return decoding_graphs
+
+    @lru_cache(maxsize=100000)
+    def compile_one_and_cache(self, text: torch.Tensor) -> k2.Fsa:
+        label_graph = k2.linear_fsa(text.tolist(), self.device)
+        label_graph = k2.add_epsilon_self_loops(label_graph)
+        decoding_graph = k2.intersect(
+            self.ctc_topo_inv, label_graph, treat_epsilons_specially=False
+        )
+        decoding_graph = k2.connect(decoding_graph.invert_())
```
[the remark this answers is not in the document]
Comment (yaguanghu) [reply espnet/nets/pytorch_backend/k2_ctc.py:141]: I tested it on the CUDA device, it seemed woking? But this operation is not necessary, I'll remove it.

File: espnet/nets/pytorch_backend/k2_ctc.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+            Tuple or tensor of size (N). It represent lengths of the targets.
+        """
+        log_probs = log_probs.permute(
+            1, 0, 2
+        )  # now log_probs is [N, T, C]  batchSize x seqLength x alphabet_size
+        supervision_segments = torch.stack(
+            (
+                torch.tensor(range(input_lengths.shape[0])),
+                torch.zeros(input_lengths.shape[0]),
+                input_lengths,
```
Comment (csukuangfj) [inline espnet/nets/pytorch_backend/k2_ctc.py:76]: Are elements of `input_lengths` sorted in descending order?

If not, I would recommend you to sort them. Please refer to
https://github.com/k2-fsa/snowfall/blob/bbcf1ec1185179b9b502287c51b46fbc2accb8d0/egs/librispeech/asr/simple_v1/ctc_train.py#L80-L81

```
    indices = torch.argsort(supervision_segments[:, 2], descending=True)
    supervision_segments = supervision_segments[indices]
```

Comment (danpovey) [reply espnet/nets/pytorch_backend/k2_ctc.py:76]: If this constraint turns out to be a hassle, we could perhaps introduce into the C++ a recursion, that checks whether they are sorted by length, and if not, sorts them by length, does the intersection and then indexes them to get them back in the correct order.
